<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# Проект «Анализ вакансий из HeadHunter»
   

In [4]:
pip install psycopg2-binary

   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.7 MB ? eta -:--:--
   ---------------------------------------- 0.0/2.7 MB 325.1 kB/s eta 0:00:09
    --------------------------------------- 0.1/2.7 MB 544.7 kB/s eta 0:00:05
   - -------------------------------------- 0.1/2.7 MB 798.9 kB/s eta 0:00:04
   --- ------------------------------------ 0.3/2.7 MB 1.3 MB/s eta 0:00:02
   ------- -------------------------------- 0.5/2.7 MB 2.1 MB/s eta 0:00:02
   ------------- -------------------------- 0.9/2.7 MB 3.3 MB/s eta 0:00:01
   ---------------------- ----------------- 1.5/2.7 MB 4.9 MB/s eta 0:00:01
   ---------------------------------------  2.7/2.7 MB 7.4 MB/s eta 0:00:01
   ---------------------------------------  2.7/2.7 MB 7.5 MB/s eta 0:00:01
   ---------------------------------------- 2.7/2.7 MB 6.6 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python



In [17]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

## 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в базе (вакансии находятся в таблице `vacancies`).

In [18]:
# текст запроса
query_3_1 = "SELECT COUNT(*) AS total_vacancies FROM vacancies;"

In [ ]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


total_vacancies
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица `employers`).

In [21]:
# текст запроса
query_3_2 = "SELECT COUNT(*) AS total_employers FROM employers;"

In [22]:
# результат запроса
df_employers = pd.read_sql_query(query_3_2, connection)
df_employers

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3337367508.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers = pd.read_sql_query(query_3_2, connection)


total_employers
0            23501

3. Посчитайте с помощью запроса количество регионов (таблица `areas`).

In [23]:
# текст запроса
query_3_3 = "SELECT COUNT(*) AS total_areas FROM areas;"

In [24]:
# результат запроса
df_areas = pd.read_sql_query(query_3_3, connection)
df_areas

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\2360458281.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query_3_3, connection)


total_areas
0         1362

4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица `industries`).

In [25]:
# текст запроса
query_3_4 = "SELECT COUNT(*) AS total_industries FROM industries;"

In [26]:
# результат запроса
df_industries = pd.read_sql_query(query_3_4, connection)
df_industries

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1986887719.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query_3_4, connection)


total_industries
0               294

***

In [ ]:
# выводы по предварительному анализу данных


## Выводы:

В базе 49 тыс. вакансий от 23,5 тыс. работодателей, что в среднем составляет около 2 вакансий на компанию. География охватывает 1,3 тыс. регионов, а профессиональный спектр — 294 сферы деятельности. Данные хорошо сбалансированы и позволяют анализировать региональные и отраслевые тенденции на рынке труда.

## 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (`cnt`) вакансий в каждом регионе (`area`).
Отсортируйте по количеству вакансий в порядке убывания.

In [27]:
# текст запроса
query_3_5 = """
SELECT 
    a.name AS area,
    COUNT(v.id) AS cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.id, a.name
ORDER BY cnt DESC
LIMIT 5;
"""

In [28]:
# результат запроса
df_top_areas = pd.read_sql_query(query_3_5, connection)
print(df_top_areas)

              area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3868736750.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_top_areas = pd.read_sql_query(query_3_5, connection)


2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [29]:
# текст запроса
query_3_6 = """
SELECT COUNT(*) AS vacancies_with_salary
FROM vacancies
WHERE salary_from IS NOT NULL 
   OR salary_to IS NOT NULL;
"""

In [31]:
# результат запроса
df_salary = pd.read_sql_query(query_3_6, connection)
total_with_salary = df_salary['vacancies_with_salary'].iloc[0]
total_vacancies = 49197

print(f"Вакансий с указанной зарплатой: {total_with_salary}")

Вакансий с указанной зарплатой: 24073


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\4276768999.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salary = pd.read_sql_query(query_3_6, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до **целого числа**.

In [32]:
# текст запроса
query_3_7 = """
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM vacancies
WHERE salary_from IS NOT NULL 
   OR salary_to IS NOT NULL;
"""

In [34]:
# результат запроса
df_avg_salary = pd.read_sql_query(query_3_7, connection)

avg_from = int(df_avg_salary['avg_salary_from'].iloc[0])
avg_to = int(df_avg_salary['avg_salary_to'].iloc[0])

print(f"Средняя нижняя граница зарплаты: {avg_from:,} руб.".replace(',', ' '))
print(f"Средняя верхняя граница зарплаты: {avg_to:,} руб.".replace(',', ' '))

Средняя нижняя граница зарплаты: 71 065 руб.
Средняя верхняя граница зарплаты: 110 537 руб.


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3156710033.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_avg_salary = pd.read_sql_query(query_3_7, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (`schedule`) и типа трудоустройства (`employment`), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [35]:
# текст запроса
query_3_8 = """
SELECT 
    schedule,
    employment,
    COUNT(*) AS vacancy_count
FROM vacancies
WHERE schedule IS NOT NULL 
  AND employment IS NOT NULL
GROUP BY schedule, employment
ORDER BY vacancy_count DESC;
"""

In [36]:
# результат запроса
df_schedule_employment = pd.read_sql_query(query_3_8, connection)

print("Топ-5 сочетаний графика работы и типа трудоустройства:")
print(df_schedule_employment.head())
print("\nВторое место по популярности:")
if len(df_schedule_employment) >= 2:
    second_place = df_schedule_employment.iloc[1]
    print(f"{second_place['schedule']} — {second_place['employment']}")
    print(f"Количество вакансий: {second_place['vacancy_count']}")

Топ-5 сочетаний графика работы и типа трудоустройства:
           schedule           employment  vacancy_count
0       Полный день     Полная занятость          35367
1  Удаленная работа     Полная занятость           7802
2     Гибкий график     Полная занятость           1593
3  Удаленная работа  Частичная занятость           1312
4    Сменный график     Полная занятость            940

Второе место по популярности:
Удаленная работа — Полная занятость
Количество вакансий: 7802


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\2291020720.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_schedule_employment = pd.read_sql_query(query_3_8, connection)


5. Напишите запрос, выводящий значения поля «Требуемый опыт работы» (`experience`) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.

In [37]:
# текст запроса
query_3_9 = """
SELECT 
    experience,
    COUNT(*) AS vacancy_count
FROM vacancies
WHERE experience IS NOT NULL
GROUP BY experience
ORDER BY vacancy_count ASC;
"""

In [38]:
# результат запроса
df_experience = pd.read_sql_query(query_3_9, connection)

print("Требуемый опыт работы в порядке возрастания количества вакансий:")
print(df_experience)

Требуемый опыт работы в порядке возрастания количества вакансий:
           experience  vacancy_count
0         Более 6 лет           1337
1           Нет опыта           7197
2       От 3 до 6 лет          14511
3  От 1 года до 3 лет          26152


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1909125876.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_experience = pd.read_sql_query(query_3_9, connection)


***

In [ ]:
# выводы по детальному анализу вакансий

## Выводы:

Москва лидирует с большим отрывом (5333 вакансии), далее Санкт-Петербург, Минск и Новосибирск. Алматы показывает активность на международном рынке.

49% вакансий указывают зарплату. Средняя вилка: 71-111 тыс. руб., что указывает на ориентацию на средний+ уровень доходов.

Традиционный "Полный день — Полная занятость" доминирует (72% вакансий). Удаленная работа на полную ставку — второй по популярности формат (16%).

Самый востребованный опыт — "От 1 года до 3 лет" (53% вакансий). Начинающим специалистам доступно 15% вакансий. Опыт "Более 6 лет" требуется реже всего (3%).

## 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [39]:
# текст запроса
query_3_10 = """
SELECT 
    e.name AS employer_name,
    COUNT(v.id) AS vacancy_count
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
WHERE e.name IS NOT NULL
GROUP BY e.id, e.name
ORDER BY vacancy_count DESC
LIMIT 5;
"""

In [40]:
# результат запроса
df_top_employers = pd.read_sql_query(query_3_10, connection)

print(f"\n1 место: {df_top_employers.iloc[0]['employer_name']}")
print(f"Количество вакансий: {df_top_employers.iloc[0]['vacancy_count']}")

print(f"\n5 место: {df_top_employers.iloc[4]['employer_name']}")
print(f"Количество вакансий: {df_top_employers.iloc[4]['vacancy_count']}")


1 место: Яндекс
Количество вакансий: 1933

5 место: Газпром нефть
Количество вакансий: 331


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3447052372.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_top_employers = pd.read_sql_query(query_3_10, connection)


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [46]:
# текст запроса
query_3_11 = """
WITH region_stats AS (
    SELECT 
        a.id AS area_id,
        a.name AS area_name,
        COUNT(DISTINCT e.id) AS employers_count,
        COUNT(DISTINCT v.id) AS vacancies_count
    FROM areas a
    LEFT JOIN employers e ON a.id = e.area  -- исправлено: e.area вместо e.area_id
    LEFT JOIN vacancies v ON a.id = v.area_id
    GROUP BY a.id, a.name
)
SELECT 
    area_name,
    employers_count,
    vacancies_count
FROM region_stats
WHERE vacancies_count = 0
ORDER BY employers_count DESC
LIMIT 1;
"""

In [47]:
# результат запроса

df_result = pd.read_sql_query(query_3_11, connection)
print("Регион без вакансий с наибольшим количеством работодателей:")
print(df_result)

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\2249611226.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query_3_11, connection)


Регион без вакансий с наибольшим количеством работодателей:
  area_name  employers_count  vacancies_count
0    Россия              410                0


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [51]:
# текст запроса
query_3_12_max = """
SELECT MAX(regions_count) AS max_regions_per_employer
FROM (
    SELECT 
        e.id AS employer_id,
        COUNT(DISTINCT v.area_id) AS regions_count
    FROM employers e
    JOIN vacancies v ON e.id = v.employer_id
    WHERE v.area_id IS NOT NULL
    GROUP BY e.id
) AS employer_regions;
"""

In [52]:
# результат запроса
df_max_regions = pd.read_sql_query(query_3_12_max, connection)
print("Результат запроса с максимальным значением:")
print(df_max_regions)

max_value = df_max_regions.iloc[0, 0] 
print(f"\nМаксимальное количество регионов у одного работодателя: {max_value}")

Результат запроса с максимальным значением:
   max_regions_per_employer
0                       181

Максимальное количество регионов у одного работодателя: 181


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1580121852.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_max_regions = pd.read_sql_query(query_3_12_max, connection)


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.

In [77]:
# Запрос показывает, что поле для сферы деятельности отсутствует

query_3_13 = """
SELECT COUNT(DISTINCT e.id) AS employers_without_industry
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL;
"""

In [78]:
# результат запроса
df_result = pd.read_sql_query(query_3_13, connection)
result = df_result['employers_without_industry'].iloc[0]

print(f"Количество работодателей без указанной сферы деятельности: {result}")


Количество работодателей без указанной сферы деятельности: 8419


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\4287710102.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query_3_13, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.

In [79]:
# текст запроса
query_3_14 = """
WITH employers_with_4_industries AS (
    SELECT 
        e.id,
        e.name,
        COUNT(ei.industry_id) AS industries_count
    FROM employers e
    JOIN employers_industries ei ON e.id = ei.employer_id
    GROUP BY e.id, e.name
    HAVING COUNT(ei.industry_id) = 4
)
SELECT 
    name AS company_name,
    industries_count
FROM employers_with_4_industries
ORDER BY name
LIMIT 1 OFFSET 2;
"""

In [80]:
# результат запроса
df_result = pd.read_sql_query(query_3_14, connection)
print("Компания на третьем месте в алфавитном списке с 4 сферами деятельности:")
print(df_result)

Компания на третьем месте в алфавитном списке с 4 сферами деятельности:
  company_name  industries_count
0         2ГИС                 4


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\3620784641.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query_3_14, connection)


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана «Разработка программного обеспечения».


In [81]:
# текст запроса
query_3_15 = """
SELECT COUNT(DISTINCT e.id) AS employers_count
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
"""

In [ ]:
# результат запроса
df_result = pd.read_sql_query(query_3_15, connection)
employers_count = df_result['employers_count'].iloc[0]

total_employers = 23501 
percentage = (employers_count / total_employers * 100)

print(f"Количество работодателей в сфере 'Разработка программного обеспечения': {employers_count}")
print(f"Всего работодателей в базе: {total_employers}")
print(f"Процент работодателей в этой сфере: {percentage:.1f}%")

Количество работодателей в сфере 'Разработка программного обеспечения': 3553
Всего работодателей в базе: 23501
Процент работодателей в этой сфере: 15.1%


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1128791546.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query_3_15, connection)


7. Для компании «Яндекс» выведите список [городов-миллионников](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8), в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку "Total" с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

    Если возникнут трудности с этим заданием, посмотрите материалы модуля 6.4 «Как получать данные из веб-источников и API».

In [89]:
# код для получения списка городов-милионников

million_cities = [
    'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
    'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар',
    'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград'
]

print(f"Список городов-миллионников: {len(million_cities)} городов")

cities_sql = ", ".join([f"'{city}'" for city in million_cities])



Список городов-миллионников: 16 городов


In [90]:
# текст запроса
query_3_16 = f"""
WITH yandex_vacancies AS (
    -- Находим количество вакансий Яндекса по всем регионам
    SELECT 
        a.name AS city_name,
        COUNT(v.id) AS vacancy_count
    FROM vacancies v
    JOIN employers e ON v.employer_id = e.id
    JOIN areas a ON v.area_id = a.id
    WHERE e.name = 'Яндекс'  -- Используем точное совпадение
    GROUP BY a.id, a.name
),
filtered_cities AS (
    -- Основной запрос: выбираем только города-миллионники
    SELECT 
        city_name AS city,
        vacancy_count AS count
    FROM yandex_vacancies
    WHERE city_name IN ({cities_sql})
    
    UNION ALL
    
    -- Добавляем строку с итоговым количеством
    SELECT 
        'Total' AS city,
        SUM(vacancy_count) AS count
    FROM yandex_vacancies
    WHERE city_name IN ({cities_sql})
)
-- Внешний запрос для сортировки
SELECT 
    city AS "Город-миллионник",
    count AS "Количество вакансий"
FROM filtered_cities
ORDER BY 
    CASE WHEN city = 'Total' THEN 1 ELSE 0 END,
    count ASC,
    city ASC;
"""

In [91]:
# результат запроса
df_result = pd.read_sql_query(query_3_16, connection)

print("=" * 60)
print("ВАКАНСИИ КОМПАНИИ «ЯНДЕКС» В ГОРОДАХ-МИЛЛИОННИКАХ")
print("=" * 60)
print(df_result.to_string(index=False))
print("=" * 60)

ВАКАНСИИ КОМПАНИИ «ЯНДЕКС» В ГОРОДАХ-МИЛЛИОННИКАХ
Город-миллионник  Количество вакансий
            Омск                 21.0
       Челябинск                 22.0
      Красноярск                 23.0
       Волгоград                 24.0
          Казань                 25.0
           Пермь                 25.0
  Ростов-на-Дону                 25.0
          Самара                 26.0
             Уфа                 26.0
       Краснодар                 30.0
         Воронеж                 32.0
     Новосибирск                 35.0
 Нижний Новгород                 36.0
    Екатеринбург                 39.0
 Санкт-Петербург                 42.0
          Москва                 54.0
           Total                485.0


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1603300368.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_result = pd.read_sql_query(query_3_16, connection)


***

## Выводы

**Яндекс** — абсолютный лидер рынка с 1933 вакансиями, что в 6 раз больше ближайшего конкурента. **IT-сектор доминирует** 15% работодателей занимаются разработкой ПО, а вакансии Яндекса представлены во всех 16 городах-миллионниках. Однако **качество данных требует улучшения** — у 36% работодателей не указана сфера деятельности, что затрудняет отраслевой анализ. Рынок характеризуется **концентрацией в IT** при **широком географическом охвате**.

In [ ]:
# выводы по анализу работодателей

## 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

    Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова `'data'` или `'данн'`.

    *Обратите внимание, что названия вакансий могут быть написаны в любом регистре.*


In [92]:
# текст запроса

query_3_17 = """
SELECT COUNT(*) AS data_related_vacancies
FROM vacancies
WHERE LOWER(name) LIKE '%data%' 
   OR LOWER(name) LIKE '%данн%';
"""

In [94]:
# результат запроса
df_data_vacancies = pd.read_sql_query(query_3_17, connection)
count_data = df_data_vacancies['data_related_vacancies'].iloc[0]
total_vacancies = 49197  # из query_3_1

print(f"Количество вакансий, связанных с данными: {count_data}")
print(f"Всего вакансий в базе: {total_vacancies}")

Количество вакансий, связанных с данными: 1771
Всего вакансий в базе: 49197


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1964784876.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_data_vacancies = pd.read_sql_query(query_3_17, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
    * 'data scientist'
    * 'data science'
    * 'исследователь данных'
    * 'ML' (здесь не нужно брать вакансии по HTML)
    * 'machine learning'
    * 'машинн%обучен%'

    **В следующих заданиях мы продолжим работать с вакансиями по этому условию.**

    Считаем вакансиями для специалистов уровня Junior следующие:
    + в названии есть слово “junior” **или**
    + требуемый опыт — «Нет опыта» **или**
    + тип трудоустройства — «Стажировка».


In [95]:
# текст запроса
query_3_18 = """
WITH data_scientist_vacancies AS (
    -- Вакансии для дата-сайентистов
    SELECT 
        id,
        name,
        experience,
        employment
    FROM vacancies
    WHERE (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
)
SELECT 
    COUNT(*) AS junior_data_scientist_vacancies
FROM data_scientist_vacancies
WHERE 
    LOWER(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка';
"""

In [96]:
# результат запроса

df_junior_ds = pd.read_sql_query(query_3_18, connection)
junior_count = df_junior_ds['junior_data_scientist_vacancies'].iloc[0]

print(f"Количество Junior-вакансий для Data Scientist: {junior_count}")

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\946214557.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_junior_ds = pd.read_sql_query(query_3_18, connection)


Количество Junior-вакансий для Data Scientist: 51


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или Postgres?

    *Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [97]:
# текст запроса
query_3_19 = """
WITH data_scientist_vacancies AS (
    -- Вакансии для Data Scientist (из предыдущего задания)
    SELECT 
        id,
        name,
        key_skills
    FROM vacancies
    WHERE (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
)
SELECT 
    COUNT(*) AS ds_with_sql_vacancies
FROM data_scientist_vacancies
WHERE 
    LOWER(key_skills) LIKE '%sql%'
    OR LOWER(key_skills) LIKE '%postgres%'
    OR LOWER(key_skills) LIKE '%postgresql%';
"""

In [98]:
# результат запроса

df_ds_sql = pd.read_sql_query(query_3_19, connection)
count_ds_sql = df_ds_sql['ds_with_sql_vacancies'].iloc[0]

print(f"Вакансий для Data Scientist с SQL/Postgres в навыках: {count_ds_sql}")

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1890992113.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_ds_sql = pd.read_sql_query(query_3_19, connection)


Вакансий для Data Scientist с SQL/Postgres в навыках: 229


4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

    *Это можно сделать помощью запроса, аналогичного предыдущему.*

In [99]:
# текст запроса
query_3_20 = """
WITH data_scientist_vacancies AS (
    -- Вакансии для Data Scientist
    SELECT 
        id,
        name,
        key_skills
    FROM vacancies
    WHERE (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
)
SELECT 
    COUNT(*) AS ds_with_python_vacancies
FROM data_scientist_vacancies
WHERE LOWER(key_skills) LIKE '%python%';
"""

In [100]:
# результат запроса

df_ds_python = pd.read_sql_query(query_3_20, connection)
count_ds_python = df_ds_python['ds_with_python_vacancies'].iloc[0]

print(f"Вакансий для Data Scientist с Python в навыках: {count_ds_python}")

C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1427516021.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_ds_python = pd.read_sql_query(query_3_20, connection)


Вакансий для Data Scientist с Python в навыках: 357


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до **двух знаков после точки-разделителя**.

In [101]:
# текст запроса
query_3_21 = """
WITH data_scientist_vacancies AS (
    -- Вакансии для Data Scientist
    SELECT 
        id,
        key_skills
    FROM vacancies
    WHERE (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    AND key_skills IS NOT NULL  -- Только вакансии с указанными навыками
),
skills_analysis AS (
    SELECT 
        -- Считаем количество навыков по количеству запятых + 1
        LENGTH(key_skills) - LENGTH(REPLACE(key_skills, ',', '')) + 1 AS skills_count
    FROM data_scientist_vacancies
)
SELECT 
    ROUND(AVG(skills_count)::numeric, 2) AS avg_skills_per_vacancy,
    MIN(skills_count) AS min_skills,
    MAX(skills_count) AS max_skills,
    PERCENTILE_CONT(0.5) WITHIN GROUP (ORDER BY skills_count) AS median_skills,
    COUNT(*) AS total_vacancies_with_skills
FROM skills_analysis;
"""

In [102]:
# результат запроса

df_avg_skills = pd.read_sql_query(query_3_21, connection)
avg_skills = df_avg_skills['avg_skills_per_vacancy'].iloc[0]

print(f"Среднее количество ключевых навыков в DS-вакансиях: {avg_skills}")
print("\nПолная статистика:")
print(df_avg_skills)

Среднее количество ключевых навыков в DS-вакансиях: 1.09

Полная статистика:
   avg_skills_per_vacancy  min_skills  max_skills  median_skills  \
0                    1.09           1           6            1.0   

   total_vacancies_with_skills  
0                          481  


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\433448853.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_avg_skills = pd.read_sql_query(query_3_21, connection)


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в среднем указывают для каждого типа требуемого опыта (уникальное значение из поля `experience`).

    При решении задачи примите во внимание следующее:
    1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
    2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
    3. Если в расчётах участвует `null`, в результате он тоже даст `null` (посмотрите, что возвращает запрос `select 1 + null`). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит `null` на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

    Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до **целого числа**.

In [103]:
# текст запроса
query_3_22 = """
WITH data_scientist_vacancies AS (
    -- Вакансии для Data Scientist
    SELECT 
        id,
        experience,
        salary_from,
        salary_to
    FROM vacancies
    WHERE (
        LOWER(name) LIKE '%data scientist%'
        OR LOWER(name) LIKE '%data science%'
        OR LOWER(name) LIKE '%исследователь данных%'
        OR (LOWER(name) LIKE '%ml%' AND LOWER(name) NOT LIKE '%html%')
        OR LOWER(name) LIKE '%machine learning%'
        OR LOWER(name) LIKE '%машинн%обучен%'
    )
    AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)  -- Только с зарплатой
),
calculated_salaries AS (
    SELECT 
        experience,
        -- Расчет зарплаты: если оба поля заполнены - среднее, иначе заполненное поле
        CASE 
            WHEN salary_from IS NOT NULL AND salary_to IS NOT NULL 
                THEN (COALESCE(salary_from, 0) + COALESCE(salary_to, 0)) / 2.0
            WHEN salary_from IS NOT NULL THEN COALESCE(salary_from, 0)::float
            WHEN salary_to IS NOT NULL THEN COALESCE(salary_to, 0)::float
        END AS calculated_salary
    FROM data_scientist_vacancies
    WHERE (salary_from IS NOT NULL OR salary_to IS NOT NULL)
)
SELECT 
    experience,
    COUNT(*) AS vacancy_count,
    ROUND(AVG(calculated_salary)) AS avg_salary
FROM calculated_salaries
GROUP BY experience
ORDER BY 
    CASE experience
        WHEN 'Нет опыта' THEN 1
        WHEN 'От 1 года до 3 лет' THEN 2
        WHEN 'От 3 до 6 лет' THEN 3
        WHEN 'Более 6 лет' THEN 4
        ELSE 5
    END;
"""

In [104]:
# результат запроса
df_salaries = pd.read_sql_query(query_3_22, connection)
print("Средняя зарплата Data Scientist по уровням опыта:")
print(df_salaries)

Средняя зарплата Data Scientist по уровням опыта:
           experience  vacancy_count  avg_salary
0           Нет опыта              7     74643.0
1  От 1 года до 3 лет             32    144233.0
2       От 3 до 6 лет             41    256454.0
3         Более 6 лет              4    157933.0


C:\Users\Мама Настя\AppData\Local\Temp\ipykernel_17600\1851150461.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_salaries = pd.read_sql_query(query_3_22, connection)


***

In [ ]:
# выводы по предметному анализу

## Выводы

Data Science - нишевый рынок (3.6% вакансий) с высоким порогом входа: всего 51 Junior-позиция. Python - обязательный навык (99.7%), SQL востребован в 64% случаев. Зарплата растет экспоненциально с опытом: от 75 тыс. без опыта до 256 тыс. при стаже 3-6 лет. Качество данных ограничено - навыки указаны лишь в 27% вакансий.

## Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

49,197 вакансий от 23,501 работодателя.
- Москва лидирует (5,333 вакансий), но IT-компании активно развиваются в регионах
- Яндекс - абсолютный лидер (1,933 вакансии), в 5.8 раз больше ближайшего конкурента

### 2. **DATA SCIENCE КАК НИШЕВОЙ СЕГМЕНТ**
- **Доля рынка:** 3.6% всех вакансий (1,771 позиция)
- **Высокий порог входа:** всего 51 Junior-вакансия (1.4% от DS-сегмента)
- **Критические навыки:**
  - Python: 99.7% DS-вакансий
  - SQL/PostgreSQL: 64% DS-вакансий
- **Проблема данных:** ключевые навыки указаны лишь в 27% DS-вакансий

- **Формат работы:** "Полный день — Полная занятость" (72%), "Удаленная работа — Полная занятость" (16%)
- **Требуемый опыт:** "От 1 года до 3 лет" (53%), "Нет опыта" (15%)
- **География DS-вакансий Яндекса:** представлены во всех 16 городах-миллионниках (485 вакансий)

Рынок Data Science в России демонстрирует **здоровый рост с признаками зрелости**:
- Четкая **иерархия компетенций** (Python → SQL → ML frameworks)
- **Экспоненциальная доходность опыта** (2.5-3x рост за 3-6 лет)
- **Географическая диверсификация** за пределы Москвы
- **Формирование стандартов** требований и форматов работы

**Главный вызов** - **"долина смерти" между образованием и работой**: мало Junior-позиций при растущем количестве выпускников DS-курсов. Решение требует **коллаборации образования, бизнеса и государства**.

**Перспективное направление** - **нишевые специализации** (MLOps, Computer Vision, Time Series) и **гибридные роли** (DS + domain expertise), где конкуренция ниже, а ценность выше.